### PR TITLE
Add virtualenv support for `mlflow models docker-build`

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

Add virtualenv support for `mlflow models docker-build`.

## How is this patch tested?

mlflow models:

```bash
# Log a model to serve
python examples/xgboost/xgboost_native/train.py

# Build a docker image that serves the model
mlflow models serve -m runs:/<run_id>/model --env-manager virtualenv

# Serve the model
docker run --rm -p 5001:8080 mlflow-pyfunc-servable:latest

# Send a request
curl -X POST -d "{\"data\":[[1,2,3,4]]}" -H "Content-Type: application/json; format=pandas-split" http://localhost:5001/invocations

# The command above should print out something similar to:
# [[0.11640435457229614, 0.21829040348529816, 0.6653051972389221]]
```

sagemaker:

```
mlflow sagemaker build-and-push-container --no-push --mlflow-home . --env-manager virtualenv
```

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/docker_utils.py [hunk 1]
```diff
@@ -60,36 +85,41 @@ def _get_mlflow_install_step(dockerfile_context_dir, mlflow_home):
...
-            "mvn --batch-mode package -DskipTests && "
-            "mkdir -p /opt/java/jars && "
-            "mv /opt/mlflow/mlflow/java/scoring/target/"
-            "mlflow-scoring-*-with-dependencies.jar /opt/java/jars\n"
+            # Temporarily commented out for faster development
+            # "RUN cd /opt/mlflow/mlflow/java/scoring && "
+            # "mvn --batch-mode package -DskipTests && "
+            # "mkdir -p /opt/java/jars && "
+            # "mv /opt/mlflow/mlflow/java/scoring/target/"
+            # "mlflow-scoring-*-with-dependencies.jar /opt/java/jars\n"
```
Comment (harupy) [inline mlflow/models/docker_utils.py:93]:
```suggestion
            "RUN cd /opt/mlflow/mlflow/java/scoring && "
            "mvn --batch-mode package -DskipTests && "
            "mkdir -p /opt/java/jars && "
            "mv /opt/mlflow/mlflow/java/scoring/target/"
            "mlflow-scoring-*-with-dependencies.jar /opt/java/jars\n"
```

File: mlflow/models/docker_utils.py [hunk 2]
```diff
@@ -60,36 +85,41 @@ def _get_mlflow_install_step(dockerfile_context_dir, mlflow_home):
...
+            # " --batch-mode dependency:copy"
+            # " -Dartifact=org.mlflow:mlflow-scoring:{version}:pom"
+            # " -DoutputDirectory=/opt/java\n"
+            # "RUN mvn "
+            # " --batch-mode dependency:copy"
+            # " -Dartifact=org.mlflow:mlflow-scoring:{version}:jar"
+            # " -DoutputDirectory=/opt/java/jars\n"
+            # "RUN cp /opt/java/mlflow-scoring-{version}.pom /opt/java/pom.xml\n"
+            # "RUN cd /opt/java && mvn "
+            # "--batch-mode dependency:copy-dependencies -DoutputDirectory=/opt/java/jars\n"
```
Comment (harupy) [inline mlflow/models/docker_utils.py:109]:
```suggestion
            "RUN mvn "
            " --batch-mode dependency:copy"
            " -Dartifact=org.mlflow:mlflow-scoring:{version}:pom"
            " -DoutputDirectory=/opt/java\n"
            "RUN mvn "
            " --batch-mode dependency:copy"
            " -Dartifact=org.mlflow:mlflow-scoring:{version}:jar"
            " -DoutputDirectory=/opt/java/jars\n"
            "RUN cp /opt/java/mlflow-scoring-{version}.pom /opt/java/pom.xml\n"
            "RUN cd /opt/java && mvn "
            "--batch-mode dependency:copy-dependencies -DoutputDirectory=/opt/java/jars\n"
```

File: mlflow/models/docker_utils.py [hunk 3]
```diff
@@ -98,13 +128,20 @@ def _build_image(image_name, entrypoint, mlflow_home=None, custom_setup_steps_ho
...
+
     with TempDir() as tmp:
         cwd = tmp.path()
         install_mlflow = _get_mlflow_install_step(cwd, mlflow_home)
         custom_setup_steps = custom_setup_steps_hook(cwd) if custom_setup_steps_hook else ""
         with open(os.path.join(cwd, "Dockerfile"), "w") as f:
             f.write(
                 _DOCKERFILE_TEMPLATE.format(
+                    setup_miniconda=setup_miniconda,
+                    setup_pyenv=setup_virtualenv,
```
Comment (harupy) [inline mlflow/models/docker_utils.py:144]:
```suggestion
                    setup_virtualenv=setup_virtualenv,
```

File: mlflow/models/container/__init__.py [hunk 4]
```diff
@@ -142,8 +151,15 @@ def _serve_pyfunc(model):
...
-        bash_cmds += ["source /miniconda/bin/activate custom_env"]
+            activate_cmd = _install_pyfunc_deps(
+                MODEL_PATH,
+                install_mlflow=True,
+                enable_mlserver=enable_mlserver,
+                env_manager=env_manager,
+            )
+        else:
+            activate_cmd = []
+        bash_cmds += activate_cmd
```
Comment (harupy) [inline mlflow/models/container/__init__.py:162]: This block needs to be fixed. When running the image, `disable_env_creation` is False so we fall into the else branch and `activate_cmd` is set to an empty array. If it's an empty array, the environment to serve the model is not activated when running the image.

File: mlflow/models/docker_utils.py [hunk 5]
```diff
@@ -6,16 +6,42 @@
...
 _logger = logging.getLogger(__name__)
 
+SETUP_MINICONDA = """
+# Setup miniconda
+RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh >> miniconda.sh
+RUN bash ./miniconda.sh -b -p /miniconda && rm ./miniconda.sh
+ENV PATH="/miniconda/bin:$PATH"
+"""
+
+SETUP_VIRTUALENV = """
```
Comment (harupy) [inline mlflow/models/docker_utils.py:20]:
```suggestion
SETUP_PYENV_AND_VIRTUALENV = """
```

File: mlflow/models/cli.py [hunk 6]
```diff
@@ -155,9 +155,10 @@ def prepare_env(
 @commands.command("build-docker")
 @cli_args.MODEL_URI
 @click.option("--name", "-n", default="mlflow-pyfunc-servable", help="Name to use for built image")
+@cli_args.ENV_MANAGER
 @cli_args.INSTALL_MLFLOW
```
Comment (dbczumar) [inline mlflow/models/cli.py:159]: Can we update the docstring for `INSTALL_MLFLOW` to indicate that MLflow will be installed into the virtualenv as well? Right now, it just mentions conda:

```
  --install-mlflow     If specified and there is a conda
                       environment to be activated mlflow will
                       be installed into the environment after
                       it has been activated. The version of
                       installed mlflow will be the same asthe
                       one used to invoke this command.
```

File: mlflow/sagemaker/cli.py [hunk 7]
```diff
@@ -567,22 +568,26 @@ def run_local(model_uri, port, image, flavor):
 @click.option("--build/--no-build", default=True, help="Build the container if set.")
 @click.option("--push/--no-push", default=True, help="Push the container to AWS ECR if set.")
 @click.option("--container", "-c", default=IMAGE, help="image name")
+@cli_args.ENV_MANAGER
 @cli_args.MLFLOW_HOME
-def build_and_push_container(build, push, container, mlflow_home):
+def build_and_push_container(build, push, container, env_manager, mlflow_home):
```
Comment (dbczumar) [inline mlflow/sagemaker/cli.py:573]: Can we add tests for this method?

File: mlflow/models/cli.py [hunk 8]
```diff
@@ -155,9 +155,10 @@ def prepare_env(
...
 @cli_args.ENABLE_MLSERVER
-def build_docker(model_uri, name, install_mlflow, enable_mlserver):
+def build_docker(model_uri, name, env_manager, install_mlflow, enable_mlserver):
```
Comment (dbczumar) [inline mlflow/models/cli.py:161]: Can we also add the `@cli_args.MLFLOW_HOME` arg to this method?

**Review by dbczumar — APPROVED** [review]
LGTM once remaining comments are addressed. Thanks @harupy !

File: mlflow/models/cli.py (hunk 8 above)
Comment (harupy) [reply mlflow/models/cli.py:161]: Thanks for the catch, sure!

File: mlflow/sagemaker/cli.py (hunk 7 above)
Comment (harupy) [reply mlflow/sagemaker/cli.py:573]: Added a test!

File: mlflow/models/cli.py (hunk 6 above)
Comment (harupy) [reply mlflow/models/cli.py:159]: Updated!

**Pull request merged**